### PR TITLE
Basic reverse-proxy handler

### PR DESCRIPTION
**Pull request by KyCodeHuynh**
This adds basic reverse-proxy functionality (with more advanced functionality to follow in a subsequent PR).

* When the `/reverse_proxy` route is visited, we look at
  at the reverse-proxy config block to see the
`remote_host` to reach and `remote_port` to use.

* We examine the request from the original client,
  and construct a new one to send to the `remote_host`.

* We parse the `remote_response` sent back by the `remote_host`
  to check its ResponseCode.

* We return the body of the `remote_response`, with the
  headers modified to have the apropriate response code.
  For example, a 500 error on the `remote_host` corresponds to
  a 404 error from the reverse-proxy, as the reverse-proxy
  itself was working well.

TODOs remaining:

* Handle 302 redirects (where we attempt to fetch the
  redirected-to content before returning it to the original
  client)

* Rewrite the URI prefixes in the body of the remote_response
  to handle HTML elements that have relative URLs (which would
  otherwise be broken by trying to reach "`reverse_proxy/styles.cc`")

### CODE REVIEW

**Comment by gotitmemorize** [comment]
I have a comment for:
- Rewrite the URI prefixes in the body of the remote_response
to handle HTML elements that have relative URLs (which would
otherwise be broken by trying to reach "reverse_proxy/styles.cc")

Isn't the response from ucla.edu compressed using gzip? which means you would have uncompress it and change it.

**Comment by Wizzard033** [comment]
Looks good so far. The style is appropriate.
I didn't test it, but I will once the final commit is in.

**Comment by KyCodeHuynh** [comment]
@UCLA-CS130/team05: To clarify, this PR is meant to be mergeable as-is, with the 302-handling and relative link-rewriting to follow in a separate PR. It does still work with the latest multi-threading, as seen in the [latest CI run](https://travis-ci.org/UCLA-CS130/Team05/builds/206824758).

Is this good-to-merge now?

**Comment by Wizzard033** [comment]
We can merge it, but I'm not gonna run and test it until it's complete.
If any of my other teammates reads this, go ahead and merge it if you're OK with that.

File: reverse_proxy_handler.cc [hunk 1]
```diff
@@ -1,22 +1,185 @@
...
+    std::cerr << "Sending remote_request: " << remote_request << std::endl;
+
+    boost::asio::write(socket, boost::asio::buffer(remote_request));
 
+    // Based on: https://github.com/UCLA-CS130/Mr.-Robot-et-al./blob/c9b064c68cd4bc1ae6b5c012db59eae9cb8b946d/lightning_server.cc#L26
+    const int MAX_BUFFER_LENGTH = 8192;
+    char response_buffer[MAX_BUFFER_LENGTH];
+    memset(response_buffer, 0, MAX_BUFFER_LENGTH);
 
+    socket.read_some(boost::asio::buffer(response_buffer), ec);
```
Comment (gotitmemorize) [inline reverse_proxy_handler.cc]: may want to put a loop here eventually for when responses exceed 8192 (for instance, when retrieving an image)

**Pull request merged**